pelita/game.py

```diff
@@ -624,25 +623,34 @@ def apply_move(gamestate, bot_position):
 
     # previous errors
     team_errors = gamestate["errors"][team]
-    # check is step is legal
+
+    # the allowed moves for the current bot
     legal_moves = get_legal_moves(walls, gamestate["bots"][gamestate["turn"]])
-    if bot_position not in legal_moves:
-        bad_bot_position = bot_position
-        bot_position = legal_moves[gamestate['rnd'].randint(0, len(legal_moves)-1)]
-        error_dict = {
-            "turn": turn,
-            "round": n_round,
-            "reason": 'illegal move',
-            "bot_position": bot_position
-            }
-        game_print(turn, f"Illegal move {bad_bot_position} not in {sorted(legal_moves)}. Choosing a random move instead: {bot_position}")
-        team_errors.append(error_dict)
+
+    # unless we have already made an error, check if we made a legal move
+    if not (n_round, turn) in team_errors:
+        if bot_position not in legal_moves:
+            error_dict = {
+                "reason": 'illegal move',
+                "bot_position": bot_position
+                }
+            # add the error to the team’s errors
+            game_print(turn, f"Illegal move {bot_position} not in {sorted(legal_moves)}.")
+            team_errors[(n_round, turn)] = error_dict
 
     # only execute move if errors not exceeded
     gamestate.update(check_gameover(gamestate))
     if gamestate['gameover']:
         return gamestate
 
+    # Now check if we must make a random move
+    if (n_round, turn) in team_errors:
+        # There was an error for this round and turn
+        # (but not enough that the game is over)
+        # We execute a random move
+        bot_position = gamestate['rnd'].choice(legal_moves)
+        game_print(turn, f"Choosing a random move instead: {bot_position}")
```

Comment:
as above, `position` instead of `move`. No, actually even better would be:
`f"Setting a random position instead: {bot_position}"`